### DISCUSSION TITLE
Autocompletion expands recursive unions

### DISCUSSION THREAD

**Discussion by qexat** [microsoft/pylance-release#6230] [Q&A] [answered]
Hi, I have a union type that is defined using the `type` statement (PEP 695).
I have a `Protocol` class which methods return this type. The autocompletion automatically expands the union. How do I prevent this?

## Example

```py
class MyProtocol(typing.Protocol):
    def method(self, arg: int) -> MyType: ...

class A[T: MyType]:
    def __init__(self, value: T) -> None:
        pass

class B:
    pass

type MyType = A[MyType] | B  # for example, not my actual case

class MyClass(MyProtocol):
    def met│
         # ^ cursor - tab is pressed here
```

### Expected

```py
class MyClass(MyProtocol):
    def method(self, arg: int) -> MyType:
        return super().method(arg)│
```

### Actual

```py
class MyClass(MyProtocol):
    def method(self, arg: int) -> A[... | B] | B:
        return super().method(arg)│
```

(imports omitted for brevity)

## Configuration

Pylance v2024.7.1

**Comment by rchiodo** [comment, marked as answer]
I think this is an issue we already have:
https://github.com/microsoft/pylance-release/issues/5464

Which was turned into this discussion item:
https://github.com/microsoft/pylance-release/discussions/6172

**Reply by qexat** [reply]
> I think this is an issue we already have: #5464
> 
> Which was turned into this discussion item: #6172

My bad. I searched for words like `union`, `type` and `autocompletion` but I somehow missed that, apologies.
